lib/src/collection/lru_map.dart

```diff
@@ -0,0 +1,286 @@
+part of quiver.collection;
+
+/**
+ * An implementation of a [Map] which has a maximum size and uses a Least
+ * Recently Used algorithm to remove items from the [Map] when the [maximumSize]
+ * is reached and new items are added.
+ */
+abstract class LruMap<K, V> implements Map<K, V> {
+  /**
+   * Creates a [LruMap] instance with the default implementation.
+   */
+  factory LruMap({int maximumSize}) = LinkedLruHashMap;
+
+  /**
+   * Maximum size of the [Map]. If [length] exceeds this value at any time, the
+   * oldest entries are removed beneath the current [length].
+   */
+  int maximumSize;
+}
+
+/**
+ * Simple implementation of a linked-list entry that contains a [key] & [value].
+ */
+class _LinkedEntry<K, V> {
+  K key;
+  V value;
+
+  _LinkedEntry<K, V> next;
+  _LinkedEntry<K, V> previous;
+
+  _LinkedEntry([this.key, this.value]);
+}
+
+/**
+ * An [Iterator]/[Iterable] combo for navigating through [_LinkedEntry].
+ */
+class _LinkedEntryIterator<K, V>
+    extends IterableBase<_LinkedEntry<K, V>>
+    implements Iterator<_LinkedEntry<K, V>> {
+  @override
+  _LinkedEntry<K, V> current;
+
+  _LinkedEntryIterator(_LinkedEntry<K, V> head) {
+    current = new _LinkedEntry<K, V>()..next = head;
+  }
+
+  @override
+  Iterator<_LinkedEntry<K, V>> get iterator => this;
+
+  @override
+  bool moveNext() {
+    if (current.next == null) return false;
+    current = current.next;
+    return true;
+  }
+}
+
+/**
+ * A linked hash-table based implementation of [LruMap].
+ */
+class LinkedLruHashMap<K, V> implements LruMap<K, V> {
+  static const _DEFAULT_MAXIMUM_SIZE = 100;
+
+  final Map<K, _LinkedEntry<K, V>> _entries;
+
+  int _maximumSize;
+
+  _LinkedEntry<K, V> _head;
+  _LinkedEntry<K, V> _tail;
+
+  /**
+   * Create a new LinkedLruHashMap with a [maximumSize].
+   */
+  factory LinkedLruHashMap({int maximumSize}) =>
+      new LinkedLruHashMap._fromMap(new HashMap<K, _LinkedEntry<K, V>>(),
+          maximumSize: maximumSize);
+
+  LinkedLruHashMap._fromMap(
+      this._entries, {
+      int maximumSize})
+          // This pattern is used instead of a default value because we want to
+          // be able to respect null values coming in from MapCache.lru.
+          : _maximumSize = maximumSize != null ? maximumSize
+              : _DEFAULT_MAXIMUM_SIZE;
+
+  /**
+   * Iterates over [other], adding all key-value pairs to the map. The MRU/LRU
+   * order is based off of [other.forEach] implementation - the first item would
+   * be the LRU, and the last item, the MRU.
```

Comment:
The first item is only the LRU of the new keys, existing keys not overwritten by `other` are more least recently used.